camayoc/qcs_models.py

```diff
@@ -42,43 +65,140 @@ def __ne__(self, other):
+    In order to create a valid host credential you must specify either a
+    password or ssh_keyfile.
+
+    Example::
+        >>> from camayoc import api
+        >>> from camayoc.qcs_models import HostCredential
+        >>> client = api.QCSClient()
+        >>> cred = HostCredential(password='foo')
+        >>> # The create method automatically sets the credential's `_id`
+        >>> cred.create()
+        >>> actual_cred = cred.read_self().json()
+        >>> # the __eq__() method is customized to
+        >>> # allow == comparison between the json returned by the server
+        >>> # and the HostCredential object used to create it.
+        >>> assert actual_cred == cred
+    """
 
     def __init__(
             self,
+            client=None,
             name=None,
             username=None,
             password=None,
             ssh_keyfile=None,
             sudo_password=None,
             _id=None):
-        """Host Credential objects encapsulate host credential data.
-
-        Host credentials can be created by instantiating a HostCredential
-        object. A unique name and username are provided by default.
-        In order to create a valid host credential you must specify either a
-        password or ssh_keyfile.
-
-        Example::
-            >>> from camayoc import api
-            >>> from camayoc.qcs_models import HostCredential
-            >>> client = api.QCSClient()
-            >>> host = HostCredential(password='foo')
-            >>> create_response = client.create_credential(host.payload())
-            >>> host._id = create_response.json()['id']
+        """Create a host credential with given data.
+
+        If no arguments are passed, then a api.Client will be initialized and a
+        uuid4 generated for the name and username.
+
+        For a HostCredential to be successfully created on the QCS server,
+        a password XOR a ssh_keyfile must be provided.
         """
-        super().__init__(name=name, _id=_id)
+        super().__init__(client=client, name=name, _id=_id)
+        self.endpoint = QCS_CREDENTIALS_PATH
         self.username = str(uuid.uuid4()) if not username else username
         self.password = password
         self.ssh_keyfile = ssh_keyfile
         self.sudo_password = sudo_password
 
     def __eq__(self, other):
-        """Return true if both objects are equal."""
+        """Return true if both objects are equal.
+
+        :param other: This can be either another HostCredential or a dictionary
+            or json object returned from the QCS server (or crafted by hand.)
+            If `other` is a HostCredential instance, the two object's fields()
+            will be compared. Otherwise, we expect the password to have been
+            masked by the server.
+        """
         if isinstance(other, HostCredential):
-            return self.__dict__ == other.__dict__
+            return self.fields() == other.fields()
```

Comment:
Would be better to raise a type error if the type of the other object is not a HostCredential